### PR TITLE
Improve retry

### PR DESCRIPTION
**Pull request by glava**
In progress. So far:
Extended FlowTaskInstance with nextDueDate and retry delay field. This now allows FlowInstance to make two decisions:
- in case task retries are exceeded to notify Executor that it failed as an instance
- in case there are still retries to notify Executor that task should be scheduled

Scheduler now checks also for dueRetriedTasks. Once the task is due it voids next due date. 

FlowInstance is in running stage until one if its tasks exceeded the number of retries. This means FlowInstance is kept in running state while tasks are in retry state. This is because there can be some tasks that are running and can be successful, and the failed task can in next run get into successful state.

### CODE REVIEW

**Comment by glava** [comment]
@adrobisch any comments so far?

I didn't like the idea of marking an instance as failed if a task fails. An instance should be in failed only if all retries are exceeded.

Right now I keep retries, retry delay and next due date on task instance.
 What now worries me that tracking this way scheduled tasks, you are not sure how many retries you had. You just had a snapshot. 

Maybe storing data into scheduler storage makes more sense.

**Comment by adrobisch** [comment]
Looks good so far! Could you elaborate on the first part? With https://github.com/flowtick/sysiphos/pull/19/files#diff-297d2a657cf9abf3395422d161594a4eR64 the instance will only be updated if one tasks exhausted all retries, which is unrecoverable, so even if at that point another task is running, its OK to set it to failed.

I think we could move the decrement of the retries to the due date update in https://github.com/flowtick/sysiphos/pull/19/files#diff-954daabaf1d90ccffbc5ca158ae7578aR23 , so we will have a fresh copy there. I dont really like the retry delay on the task instance level, in the future this would probably be a configurable thing on the flow definition?

File: akka/src/main/scala/com/flowtick/sysiphos/execution/FlowExecution.scala [hunk 1]
```diff
@@ -19,6 +20,22 @@ trait FlowExecution extends Logging {
...
+  def dueTaskRetries(now: Long): Future[Seq[FlowInstance]] = {
+    flowTaskInstanceRepository.getScheduled().flatMap { tasks =>
+      Future.sequence(tasks.map { task =>
+        task.nextDueDate match {
+          case Some(timestamp) if timestamp <= now =>
+            log.info(s"found ${task.id} that is due for retry")
+            for {
+              instance <- flowInstanceRepository.findById(task.flowInstanceId)
+              _ <- flowTaskInstanceRepository.setNextDueDate(task.id, None)
+            } yield instance
```
Comment (glava) [inline akka/src/main/scala/com/flowtick/sysiphos/execution/FlowExecution.scala]: You mean add the retries here?

Comment (adrobisch) [reply akka/src/main/scala/com/flowtick/sysiphos/execution/FlowExecution.scala]: yes

**Pull request merged**